src/command/pin.rs

```diff
@@ -3,42 +3,40 @@ use structopt::StructOpt;
 use notion_core::error::ErrorDetails;
 use notion_core::session::{ActivityKind, Session};
 use notion_core::tool::ToolSpec;
-use notion_core::version::VersionSpec;
 use notion_fail::{throw, ExitCode, Fallible};
 
 use crate::command::Command;
 
 #[derive(StructOpt)]
 pub(crate) struct Pin {
-    /// The tool to install, e.g. `node` or `npm` or `yarn`
-    tool: String,
-
-    /// The version of the tool to install, e.g. `1.2.3` or `latest`
-    version: Option<String>,
+    /// Tools to pin, like `node@lts` or `yarn@^1.14`.
+    #[structopt(
+        name = "tool[@version]",
+        required = true,
+        min_values = 1,
+        parse(try_from_str = "ToolSpec::try_from_str")
+    )]
+    tools: Vec<ToolSpec>,
 }
 
 impl Command for Pin {
-    fn run(self, session: &mut Session) -> Fallible<ExitCode> {
-        session.add_event_start(ActivityKind::Pin);
-
-        let version = match self.version {
-            Some(version_string) => VersionSpec::parse(&version_string)?,
-            None => VersionSpec::default(),
-        };
-
-        let tool = ToolSpec::from_str_and_version(&self.tool, version);
-
-        match tool {
-            ToolSpec::Node(version) => session.pin_node(&version)?,
-            ToolSpec::Yarn(version) => session.pin_yarn(&version)?,
-            // ISSUE(#292): Implement install for npm
-            ToolSpec::Npm(_version) => unimplemented!("Pinning npm is not supported yet"),
-            ToolSpec::Package(name, _version) => {
-                throw!(ErrorDetails::CannotPinPackage { package: name })
+    fn run(mut self, session: &mut Session) -> Fallible<ExitCode> {
+        self.tools.sort();
+
+        for tool in self.tools {
+            session.add_event_start(ActivityKind::Pin);
+            match tool {
+                ToolSpec::Node(version) => session.pin_node(&version)?,
```

Comment:
I think the fail early approach is reasonable for the MVP, but that long-term we might want to be more robust (that may require some changes to the error reporting to support multiple errors). We can flesh it out more in the RFC, but that shouldn't be a blocker for this implementation.

Reply:
Seems reasonable to me. I was thinking similarly. We should also in principle be able to parallelize those with something like [Rayon](https://crates.io/crates/rayon) and then capture the results in a nice way.

Reply:
Filed issue #389 for this so it doesn't fall through the cracks